### ISSUE TITLE
Bug in GetDerivative?

### ISSUE THREAD

**Issue by Watchmojo** [mfem/mfem#39]
Hi,

I tried to test the resolution capability of MFEM by calculating derivatives of sin(k*x). 7 elements are used to divide 2\pi in the x direction and 5th order H1_FECollection is employed. Here are the error when k=1,2,3 and 4. The accuracy is questionable. Is there any way to improve this? I used wrote some spectral element code, which gives much better results than this. 

![snapshot9](https://cloud.githubusercontent.com/assets/16857800/12836064/0694cac6-cb7b-11e5-98d2-53bb01f14501.png)

Thank you,
Frank

**Comment by v-dobrev** [comment]
Hello Frank,

The H1 finite elements use Gauss-Lobatto points as degrees of freedom, and therefore projecting an analytic coefficient uses the values at those points as well. As far as I know thid is the standard basis used in the spectral element method. In the case you consider, the GetDerivative method is exact as long as the target GridFunction uses high enough order (4 in your case). The only source of error is in the initial projection to the H1 space. One way to control that projection is to use a different set of points - in MFEM this can be achieved by replacing [this line](https://github.com/mfem/mfem/blob/master/fem/fe.cpp#L6214) (which generates the 1D Gauss-Lobatto points) with a call that generates another set of points in [0,1] including the end points.

In your spectral element code what type of projection did you use?

Thanks,
Veselin

**Comment by Watchmojo** [comment]
Hi, Veselin,

I did try to use GLL points by tentatively making "GaussLobattoPoints(const int p, double *x)" calculating GLL points instead. However, the result has not been improved. 

I tried to solve the poiseuille flow, which has a parabolic solution u=y^2-1. Without the convection term u_j \partial u_i/\partial x_j, this solution is obtained very accurately. However, when the convection term is introduced, it diverged. **Even if the exact solution is projected to the GridFunction, it appears that du/dx is not zero. **I then tried to filter before calculating the nonlinear term. I also tried to project the solution to 3/2order GridFunction, calculate nonlinear term and project back. Neither of these methods worked. 

In my SEM code, GLL points are used. It is also common to use GLL for velocity and GL for pressure.

Thank you,
Frank
